[RFC 0146] Meta.Categories, not Filesystem Directory Trees (#146)
NixOS/rfcs: rfcs/0146-meta-categories.md

   The current one-size-fits-all restriction is artificial, imposes unreasonable
   limitations and results in incomplete and confusing information.

   - No, symlinks or hardlinks are not convenient for this purpose; not all
     environments support them (falling on the "less features than others"
     problem expressed before) and they convey nothing besides confusion - just
     think about writing the corresponding entry in `all-packages.nix`.

5. It puts over the (possibly human) package writer the mental load of where to
   put the files on the filesystem hierarchy, deviating them from the job of
   really writing them.

   - Or just taking the shortest path and throw it on a folder under `misc`.

6. It "locks" the filesystem, preventing its usage for other, more sensible
   purposes.

7. The most important: the categorization is not discoverable via Nix language
   infrastructure.

   Indeed there is no higher level way to query about such categories besides
   the one described in the bullet 2 above.

In light of such a bunch of problems, this RFC proposes a novel alternative to
the above mess: new `meta` attributes.

> 

Reviewer: Would be nice to have some motivation why we want categories in the first place, since it will inform the discussion/implementation.

If the primary motivation was just dealing with filesystem constraints, categories will not be relevant with RFC #140, and nuking them might be a good idea.

Maybe we want to make package discovery easier. But is the effort to implement categories worth it? How often do people look for a software in a specific category? I typically look elsewhere first e.g. at [topic-specific curated lists](https://github.com/sindresorhus/awesome) and then just search Nixpkgs for a specific software project I liked.

Author: Well, categories are very traditional, many other package databases include them. As an example, in NetBSD pkgsrc it is even mandatory:

https://github.com/NetBSD/pkgsrc/blob/trunk/mk/bsd.pkg.mk#L130

And this specific directive is very old, older than NixOS itself - here is a similar code from 2022-12-31:

https://github.com/NetBSD/pkgsrc/blob/7751a84cdeb294f49bb9143d4e502a0354d75006/mk/bsd.pkg.mk#L845

Therefore, the weight of tradition is very important to keep it.

(Good! I will include this!)

----------

Personally I don't have a great idea of use case.

Maybe I am just curious and willing to test a different software from a similar category - like terminal emulators or window managers.

Maybe I want to do a curated list of a certain type of software - games and emulators I want to put on a custom ISO.

Maybe I want to make a comparison about softwares on a same category for a research project - then I can just throw a term like "debugger" or "programming language" and collect the results.

Reviewer 2: I can see a few use cases:

1. It makes searching for some stuff easier. Currently, I'm going to search.nixos.org , I enter "sqlite" and then search for a package I need. But I could just enter "sqlite" and pick category "python" to see all of the Python libraries related to sqlite. Easier than picking a query that would search for both, more accurate than prefixing the query with `python311Packages` (also one shouldn't have to remember that namespace name)
2. It makes indexing easier. Currently we only have search.nixos.org which uses some arcane sorcery to do that. If we have categories/tags, indexing is trivial with any language and any database
3. Those two factors make offline package searching sane

Author: The "feature" already exists. I am not advocating for a resurrection of anything already dead, but for a better interface for something that already exists.

And, hey, are you tired of traditions? Try to reopen RFC #122, your "it is like resurrecting FHS" argument will be very convincing! XD

Reviewer 3: I don't understand the remark on RFC112, letting this aside

You say:

> Personally I don't have a great idea of use case.

And I think that @jtojnar remark on having usecases before doing busywork is probably a good idea.

Now, it seems like people are coming up with usecases and I feel the RFC would become more interesting by writing them down and weighting them down, my point is only that tradition is not a good enough reason to include a feature.

Author: > I don't understand the remark on RFC112, letting this aside
> 
> You say:
> 
> > Personally I don't have a great idea of use case.
> 

I have ideas, I show them. I just said they aren't so great.

> And I think that @jtojnar remark on having usecases before doing busywork is probably a good idea.
>

Maybe. 

But I don't want to divert the purpose of this RFC. I will not do a huge, full-fledged, hardcore defense of categorization, given that everyone does categorizations even in the current Nixpkgs codebase.

> my point is only that tradition is not a good enough reason to include a feature.

As I said, I am not including an inexistent thing, but enhancing something already existent.

Further, I am not treating the tradition argument as an all-or-nothing argument.

And if you don't like the word "tradition", change it to "standard".

    '';
  };

  compiler = {
    name = "Compiler";
    description = ''
      A program that converts a source from a language to another, usually from
      a higher, human-readable level to a lower, machine level.
    '';
  };

  font = {
    name = "Font";
    description = ''
      A set of files that defines a set of graphically-related glyphs.
    '';
  };

  game = {
    name = "Game";
    description = ''
      A program developed with entertainment in mind.
    '';
  };

  interpreter = {
    name = "Interpreter";
    description = ''
      A program that directly executes instructions written in a programming
      language, without requiring compilation into the native machine language.
    '';
  };

```

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

In file bochs/default.nix:

```nix
stdenv.mkDerivation {

. . .

  meta = {
    . . .
    categories = with lib.categories; [ emulator debugger ];
    . . .
    };
  };
}

```

# Drawbacks
[drawbacks]: #drawbacks

The most immediate drawbacks are:

1. A huge treewide edit of Nixpkgs

   On the other hand, this is easily sprintable and amenable to automation.
   
2. Bikeshedding

>    How many and which categories we should create? Can we expand them later? 

Reviewer: We may draw on existing Freedesktop standards https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-categories

Author: Like this?

https://specifications.freedesktop.org/menu-spec/latest/apa.html

# Alternatives
[alternatives]: #alternatives

1. Do nothing

2. Ignore/nuke the categorization completely

   This is not an idea as bad as it appear. After all, categorization has a
   non-negligible propensity to bikeshedding. Removing it removes all problems.
   
   Nonetheless, other good software collections do this just fine, and we can
   easily imitate them. Indeed, we can follow/take a peek at how Repology keeps
   the categorizations defined by those software collections.

> 

Reviewer: We could also rely on metadata from Appstream metainfo files inside the packages and push for upstreaming it. Distros like [Debian](https://wiki.debian.org/AppStream) already do that to some extent. See also https://discourse.nixos.org/t/brainstorming-generating-appstream-metadata/19273.

But yeah, that will be most work intensive and long term as well as require more infrastructure so probably best combine it with other approach in the short term.

> 

Reviewer: Another alternative might be an open taxonomy.

Author: Well, I believe I can produce some preliminar code on this regard.
Basically I can verbatim copy-paste the categories of Freedesktop, just for start.

Reviewer 4: Here's a largely uncontrolled categorization, although skewed by a built in scaffolding tool that only suggests some 20 categories: [Haskell packages by category](https://hackage.haskell.org/packages/)

My takeaways:

 - listing an uncurated set of categories does not appear to be useful
    - curating and restricting are different ideas!
 - some categories highlight library ecosystems (e.g. the Scotty category) - this could alternatively be captured by a related packages attribute (presumably out of scope, but worth considering)
 - we might want to embed these categories into ours - if we go with some open categories

Author: I have proposed something like "related categories" on the preliminary code:

https://github.com/NixOS/nixpkgs/pull/230439

    '';
  };

  compiler = {
    name = "Compiler";
    description = ''
      A program that converts a source from a language to another, usually from
      a higher, human-readable level to a lower, machine level.
    '';
  };

  font = {
    name = "Font";
    description = ''
      A set of files that defines a set of graphically-related glyphs.
    '';
  };

  game = {
    name = "Game";
    description = ''
      A program developed with entertainment in mind.
    '';
  };

  interpreter = {
    name = "Interpreter";
    description = ''
      A program that directly executes instructions written in a programming
      language, without requiring compilation into the native machine language.
    '';
  };

```

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

In file bochs/default.nix:

```nix
stdenv.mkDerivation {

. . .

  meta = {
    . . .
    categories = with lib.categories; [ emulator debugger ];
    . . .
    };
  };
}

```

# Drawbacks
[drawbacks]: #drawbacks

The most immediate drawbacks are:

1. A huge treewide edit of Nixpkgs

   On the other hand, this is easily sprintable and amenable to automation.

2. Bikeshedding

>    How many and which categories we should create? Can we expand them later?

Reviewer 5: The fact that categories can be added or retired means that people can bikeshed to their hearts' content, without blocking other processes.

   The current one-size-fits-all restriction is artificial, imposes unreasonable
   limitations and results in incomplete and confusing information.

   - No, symlinks or hardlinks are not convenient for this purpose; not all
     environments support them (falling on the "less features than others"
     problem expressed before) and they convey nothing besides confusion - just
     think about writing the corresponding entry in `all-packages.nix`.

5. It puts over the (possibly human) package writer the mental load of where to
   put the files on the filesystem hierarchy, deviating them from the job of
   really writing them.

   - Or just taking the shortest path and throw it on a folder under `misc`.

6. It "locks" the filesystem, preventing its usage for other, more sensible
   purposes.

7. The most important: the categorization is not discoverable via Nix language
   infrastructure.

   Indeed there is no higher level way to query about such categories besides
   the one described in the bullet 2 above.

In light of such a bunch of problems, this RFC proposes a novel alternative to
the above mess: new `meta` attributes.

# Detailed design
[design]: #detailed-design

> A new attribute, `meta.categories`, will be included for every Nix expression

Reviewer 6: Overlays have always been there; I guess the point is that with `meta.categories` one can query categories after overlay application without separately looking at each layer?

  };

  compiler = {
    name = "Compiler";
    description = ''
      A program that converts a source from a language to another, usually from
      a higher, human-readable level to a lower, machine level.
    '';
  };

  font = {
    name = "Font";
    description = ''
      A set of files that defines a set of graphically-related glyphs.
    '';
  };

  game = {
    name = "Game";
    description = ''
      A program developed with entertainment in mind.
    '';
  };

  interpreter = {
    name = "Interpreter";
    description = ''
      A program that directly executes instructions written in a programming
      language, without requiring compilation into the native machine language.
    '';
  };

```

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

In file bochs/default.nix:

```nix
stdenv.mkDerivation {

. . .

  meta = {
    . . .
    categories = with lib.categories; [ emulator debugger ];
    . . .
    };
  };
}

```

In a `nix repl`:

```
nix-repl> :l <nixpkgs>
Added XXXXXX variables.

nix-repl> pkgs.bochs.meta.categories
[ { ... } ]

nix-repl> map (z: z.name) pkgs.bochs.meta.categories
[ "debugger" "emulator" ]
```

# Drawbacks

> [drawbacks]: #drawbacks

Reviewer 7: What about ease/speed of finding all packages of a specific category or a set of categories ?

I mean: by using a list of categories it means that to find all `compiler` packages we would have to go through every packages `categories` list, and compare every elements of that list.
There is no fast hash lookup like `meta.categories.compiler`.

(not 100% sure about it) One way to make single category lookup fast could be to generate a lookup attrset like:
```nix
meta.categories_lookup = { compiler = lib.categories.compiler; . . . };
```

Another way could be to not use a list of categories in the first place and directly use an attrset like that one above. And get the added benerit of inherit to easily write it:
```nix
meta.categories = { inherit (lib.categories) compiler; . . . };
```

To check if a package has a set of categories I don't think there's a better way than to check if the pkg is in the first category, then the second, etc.. 

What do you think?

Author: Hum, the problem I can see is that it would be not orthogonal. Look:

```nix
{
. . .
  meta = {
    maintainers = [ maintainers.MeMyselfAndI ];
    categories = { inherit (lib.categories) compiler; }
  };
. . .
}
```

Why maintainers is a list while categories is a strange dictionary?

Reviewer 8: > Why maintainers is a list while categories is a strange dictionary?

That can be solved by migrating maintainers list to a dictionary.

Author: And `meta.platfoms` too?

```
nix-repl> :l <nixpkgs>
:l <nixpkgs>
Added 16546 variables.

nix-repl> pkgs.bochs.meta.platforms
pkgs.bochs.meta.platforms
[ "i686-cygwin" "x86_64-cygwin" "x86_64-darwin" "i686-darwin" "aarch64-darwin" "armv7a-darwin" "i686-freebsd" "x86_64-freebsd" "x86_64-solaris" "aarch64-linux" "armv5tel-linux" "armv6l-linux" "armv7a-linux" "armv7l-linux" "i686-linux" "m68k-linux" "mipsel-linux" "mips64el-linux" "powerpc64-linux" "powerpc64le-linux" "riscv32-linux" "riscv64-linux" "s390-linux" "s390x-linux" "x86_64-linux" "aarch64-netbsd" "armv6l-netbsd" "armv7a-netbsd" "armv7l-netbsd" "i686-netbsd" "m68k-netbsd" "mipsel-netbsd" "powerpc-netbsd" "riscv32-netbsd" "riscv64-netbsd" "x86_64-netbsd" "i686-openbsd" "x86_64-openbsd" "x86_64-redox" ]

nix-repl> 
```

Reviewer 6: One could switch meta to attrubute sets, but currently it's all lists, yeah, so makes sense to follow this tradition

---
feature: Decouple filesystem from categorization
start-date: 2023-04-23
author: Anderson Torres
co-authors: (find a buddy later to help out with the RFC)

> shepherd-team: (names, to be nominated and accepted by RFC steering committee)
> shepherd-leader: (name to be appointed by RFC steering committee)

Reviewer 9: ```suggestion
shepherd-team: @natsukium @fgaz @infinisil 
shepherd-leader: (name to be appointed by RFC steering committee)
```

Would anyone like to be the leader?

     it completely.

   - As said before, the categorization is already present; this RFC proposes to
     expose it to a higher level, in a structured, more discoverable format.

   - Categorization is very traditional among software collections. Many of them
     are doing this just fine for years on end, and Nixpkgs can imitate them
     easily - and even surpass them, given the benefits of Nix language
     machinery.

   - Categorization is useful in many scenarios and use cases - indeed they
     are ubiquitous in software world:
     - specialized search engines (from Repology to MELPA)
     - code forges, from Sourceforge to Gitlab
     - as said above, software collections from pkgsrc to slackbuilds
     - to organization and preservation (as Software Heritage)

# Unresolved questions
[unresolved]: #unresolved-questions

Still unsolved is what data structure is better suited to represent a category.

- For now we stick to a set `{ name, description }`.
- Given the redundancy of the option above, another possibility is something
  like `nameOfCategory = { description = ""; . . . }`

# Future work
[future]: #future-work

> - Curate the categories.

Reviewer 10: This is imo the most important thing, a categorization is useless if it isn't curated. Unfortunately this is really not trivial, because both the set of packages and categories are really large, any package could be in any category, and both of them change over time.

So I would like to see some automated process for this, here's an idea for a very fancy one:
- When adding a package, one goes through a wizard-style questionnaire, answering questions about the package, ultimately determining whether each category applies, ideally as efficiently as possible.
- When making a proposal to add a new category, these steps need to be done:
  - Determine how it relates to other categories: Is it a sub-category, does it conflict with others?
  - Justify that the category is useful: Are there a bunch of packages it can be applied to?
  - Clearly defined the category, such that it's easy to determine for package maintainers whether the category applies
  - Somehow ask each package maintainer whether this new category applies to their package. This should be as simple to answer for maintainers as possible, e.g. just a single button.

I have a feeling that a Nixpkgs-specific website would be great to integrate this all together, along with E-Mails for notifications. Maybe this could be combined into e.g. a maximum once-per-week E-Mail, asking package maintainers to perform necessary actions to maintain their packages, which could also include things like reviewing relevant PR's or issues, tending to Hydra build failures, etc.

Author: 1. This whole thing will require a team
2. Would this be the return of the mail-driven update requests? I really miss the Hydra "your package failed/passed" messages on my mailbox.

Reviewer 11: I think we shouldn't reinvent the wheel and work as much as possible with existing categorization data. For example, many applications have desktop files whose categories (and tags) one could use. And maybe repology has something useful too

Reviewer 10: I guess that _is_ a benefit of having to decide on _exactly one_ category for a package: Either you'll find the most accurate category, or none of them apply which justifies creating a new one. Therefore there is no general category maintenance required, it happens naturally as packages get added.

We should consider this as an alternative, something like
```nix
meta = {
  category = lib.categories.compiler;
}
```

And if we want to be more specific we could have category and subcategory, or something like that

Author: I feel this is not good, because it would defeat the cases of multi-valued categorization.

What about mail server, game server, file server...? One category for each looks too much.

Reviewer 10: Yeah I agree it's not perfect, but if I had to choose between a curated single-valued categorization and an uncurated multi-valued one, I think I'd pick the former.

Reviewer 6: At least the multi-valued version can include all that others (including upstreams, if they provide _some of the many_ categorisation labels) think too. Agreeing on the layout of anything single-valued will go as well as any attempt to globally agree on anything in Nixpkgs, it will be a weird and inconvenient to all (but in different ways, of course) compromise.

Author: > Yeah I agree it's not perfect, but if I had to choose between a curated single-valued categorization and an uncurated multi-valued one, I think I'd pick the former.

An uncurated multi-valued can possibly be curated in a mostly independent fashion - curating does not block a package's inclusion.

On the other hand, seeking the most fitting one-valued category puts a non-negligible mental load.

---
feature: Decouple filesystem from categorization
start-date: 2023-04-23
author: Anderson Torres
co-authors:

> shepherd-team: @7c6f434c @natsukium @fgaz @infinisil

Reviewer 10: ```suggestion
shepherd-team: @7c6f434c @natsukium @fgaz
```

I'd like to retract my involvement as an RFC shepherd. After [discussing this in the previous Nixpkgs Architecture Team meeting](https://discourse.nixos.org/t/2023-10-31-nixpkgs-architecture-team-meeting-45/34846), I've concluded that it's not something I'm very interested in. My passion is mostly in solving the most prominent existing problems in Nixpkgs. This RFC however is more about introducing something new, which might also become a prominent problem if not done correctly, but for now is not.

Furthermore, the more things I'm simultaneously doing, the less focus I have, the less efficient I am, making everything just take longer in the end (also me being a bit over-perfectionistic doesn't help).

I can review this RFC again when/if it reaches FCP, but I intend to be open about the approach and not insist on anything I mentioned previously.

Reviewer 6: Thanks for the work you have done as a shepherd, and also thanks for all the arguments you have provided, regardless of their status!

Author: I feel a bit sad, however I understand your motivations.
Nonetheless, many thanks for the input. I will think about the debtags.

Reviewer 6: @AndersonTorres I think you should accept the suggested change here…

Reviewer 6: @AndersonTorres I think you need to accept this suggestion before the FCP ends, so that the final shepherd list in the repository is the list of those who have formally signed off on starting the FCP

Author: wow, I thought it was already done!

   A piece of software can fulfill many categories; e.g.
   - an educational game
   - a console emulator (vs. a PC emulator)
   - and a special-purpose programming language (say, a smart-contracts one).

   The current one-size-fits-all restriction is artificial, imposes unreasonable
   limitations and results in incomplete and confusing information.

   - No, symlinks or hardlinks are not convenient for this purpose; not all
     environments support them (falling on the "less features than others"
     problem expressed before) and they convey nothing besides confusion - just
     think about writing the corresponding entry in `all-packages.nix`.

5. It puts over the (possibly human) package writer the mental load of where to
   put the files on the filesystem hierarchy, deviating them from the job of
   really writing them.

   - Or just taking the shortest path and throw it on a folder under `misc`.

6. It "locks" the filesystem, preventing its usage for other, more sensible
   purposes.

7. The most important: the categorization is not discoverable via Nix language
   infrastructure.

   Indeed there is no higher level way to query about such categories besides

>    the one described in the bullet 2 above.

Reviewer 12: Idk, this might be a bit off topic, but one use case this RFC would enable is to have something like `nix-shell -p` but for launcher like rofi, dmenu, etc... because I would imagine that there would be a tag for graphical applications and then one can filter for those and dynamically use nix to load them like one would with a nix-shell.

Reviewer 6: I am not sure how to word such a use case carefully, given that many packages contain multiple GUI applications that can be launched without arguments, and then some more programs where command-line arguments are expected.

Author: To me it looks like an extra application that uses the Nixpkgs metadata as a filter.  Something like `nix-index` with extras as a backend for a launcher.

     easily - and even surpass them, given the benefits of Nix language
     machinery.

   - Categorization is useful in many scenarios and use cases - indeed they
     are ubiquitous in software world:
     - specialized search engines (from Repology to MELPA)
     - code forges, from Sourceforge to Gitlab
     - as said above, software collections from pkgsrc to slackbuilds
     - to organization and preservation (as Software Heritage)

# Prior art
[prior-art]: #prior-art

As said above, categorization is very traditional among software collections. It
is not hard to cite examples in this arena; the most interesting ones I have
found are listed above (linked at [references section](#references))

- FreeBSD Ports;
- Debtags;

# Unresolved questions
[unresolved]: #unresolved-questions

Still unsolved is what data structure is better suited to represent a category.

- For now we stick to a set `{ name, description }`.
- Given the redundancy of the option above, another possibility is something
  like `nameOfCategory = { description = ""; . . . }`

# Future work
[future]: #future-work

- Curate the categories.
- Update documentation.

> - Update Continuous Integration.

Reviewer 6: Agree that it would be nice, unsure if it needs spelling out explicitly as a part of «Update CI». 

Or maybe some version of this could go into motivation?

     - specialized search engines (from Repology to MELPA)
     - code forges, from Sourceforge to Gitlab
     - as said above, software collections from pkgsrc to slackbuilds
     - organization and preservation (as Software Heritage)

# Prior art
[prior-art]: #prior-art

As said above, categorization is very traditional among software collections. It
is not hard to cite examples in this arena; the most interesting ones I have
found are listed above (linked at [references section](#references))

- FreeBSD Ports;
- Debtags;

# Unresolved questions
[unresolved]: #unresolved-questions

Still unsolved is what data structure is better suited to represent a category.

- For now we stick to a set `{ name, description }`.

# Future work
[future]: #future-work

Given the typical complexities that arise from categorization, and expecting
that regular maintainers are not expected to understand its minuteness, it is
strongly recommended the creation of a team entrusted to manage issues related
to categorization, including but not limited to:

- Update documentation.
- Curate the categories.
- Update Continuous Integration.
- Integrate categorization over the current codebase.

> Such a team should receive authority to carry out their duties:

Author: OK, let's see.

     - organization and preservation (as Software Heritage)

3. Debtags/Appstream hybrid approach

A hybrid approach for code implementation would be implement two meta
attributes, namely

- `meta.categories` for Appstream-based categories
  -  the corresponding `lib.categories` should follow Appstream closely, with
     few room to custom/extra categories
- `meta.tags` for Debtags-like tags
  - while being inspired from the venerable Debtags work, the corresponding
    `lib.tags` is completely free to modify and even divert from Debtags,
    following its own way
- generally speaking, `lib.tags` should be less bureaucratic than
  `lib.categories`

However, this approach arguably elevates the complexity of the whole work, and
adds too much redundancy.

# Prior art
[prior-art]: #prior-art

As said above, categorization is very traditional among software collections. It
is not hard to cite examples in this arena; the most interesting ones I have
found are listed below (linked at [references section](#references)):

- FreeBSD Ports;
- Debtags;
- Appstream Project;

# Unresolved questions
[unresolved]: #unresolved-questions

> There are some still unsolved issues:

Reviewer 6: ```suggestion
There are some still unsolved issues for the categorisation team to resolve:
```

Author: Done.

- What data structure is suitable to represent a category?

  - For now we stick to the most natural: a set `{ name, description }`.

- Should we have a set of primary, "most important" categories with mandatory
  status, in the sense each package should set at least one of them?

>   - The answer is most certainly positive.

Reviewer 13: I'd appreciate it if the RFC made it clear if it expects these questions to be answered by the categorization team or if it reserves them for future RFCs.

My personal take is the first question should be left to the team and the second should be answered by this RFC.

Reviewer 6: Arguably, the current text of the entire RFC gives the team authority to resolve both as possibilities of implementation, but puts a strong preference on the choice in the second one.

What exactly does it mean «each package» for primary categories in some corner cases will probably have to be refined iteratively, though, so the team gets some freedom on refining the exact interpretation without going against the letter of the RFC.

I agree that it is better to spell out that the issues are for the team.

Author: Why not both?
The categorization team decides whether it can be solved right away or if it deserves another RFC.

```

### Hybrid approach
[hybrid-approach]: #hybrid-approach

A hybrid approach for code implementation would be implement two meta
attributes, namely

- `meta.categories` for Appstream-based categories
  -  the corresponding `lib.categories` should follow Appstream closely, with
     few room to custom/extra categories

> - `meta.tags` for Debtags-like tags

Reviewer 6: I think we have discussed that as categories are sometimes a list anyway, «tags» can just got into the end of the same list?

Author: I will move this to alternatives.

# Future work
[future]: #future-work

Given the typical complexities that arise from categorization, and expecting
that regular maintainers are not expected to understand its minuteness, it is
strongly recommended the creation of a team entrusted to manage issues related

> to categorization, including but not limited to:

Reviewer 6: I want to be clear that some decisions by a team carry the authority granted by an RFC. So, a bit of RFC46-style, work→authority style proposal:

… it is strongly recommended to create a team that would take up the work of maintaining and curating the categorisation, as well as the corresponding documentation. A team doing this work will automatically receive the additional authority to:

- Resolve the disputes over categorisation of specific packages.
- Decide when a Continuous Integration check for categorisation is ready to be developed with «gray/neutral» failure statuses, and when a CI check with a good track record in «gray» mode can be upgraded to «red/blocking» failures.
- Coordinate the treewide efforts to import or update the categorisation of packages.

# Future work
[future]: #future-work

> - Create a team to manage categorization-related issues, including but not

Reviewer 6: I think this needs to be in the body as we are explicitly confirming the authority of the team to make some global-impact decisions (like when CI deployment is good enough for green-red instead of green-gray)?

Author: Done.

found are listed below (linked at [references section](#references)):

- FreeBSD Ports;
- Debtags;
- Appstream Project;

# Unresolved questions
[unresolved]: #unresolved-questions

There are some still unsolved issues:

- What data structure is suitable to represent a category?

  - For now we stick to the most natural: a set `{ name, description }`.

- Should we have a set of primary, "most important" categories with mandatory
  status, in the sense each package should set at least one of them?
  - The answer is most certainly positive.

# Future work
[future]: #future-work

- Create the [categorization team](#categorization-team)
- Carry out the duties correlated to categorization, including but not limited
  to:

  - Decide between possibilities of implementation;
  - Documentation updates;
  - Category curation, integration and updates;
  - Continuous Integration updates and adaptations;
  - Coordinaton of efforts to import, integrate and update categorization of
    packages;
  - Litigations and disputations:
    - Solve disputations over categorization, especially corner cases;
    - Solve and enforce implementation issues from CI checks

>       - When should a CI check about categorization be treated as blocking?

Reviewer 6: ```suggestion
      - When should a CI check about categorization be converted to blocking?
```
(to signal that adding blocking checks directly is rarely a good idea)

Author: OK.
